test/framework/exec/docker.go

```diff
@@ -0,0 +1,360 @@
+	// Read out any output from the call
+	outputErrors := make(chan error)
+	go func() {
+		// Send the output to the host file
+		_, err = stdcopy.StdCopy(fileOnHost, fileOnHost, resp.Reader)
+		outputErrors <- err
+	}()
+
+	select {
+	case err := <-outputErrors:
+		if err != nil {
+			return err
+		}
+
+	case <-ctx.Done():
+		return err
+	}
+
+	return nil
+}
+
+// RunContainer will run a docker container with the given settings and arguments, returing any errors.
+func RunContainer(ctx context.Context, image, net, user, group string, volumes, env map[string]string, output io.Writer, args ...string) error {
+	docker, err := getDockerClient()
+	if err != nil {
+		return err
+	}
+
+	bindings := []string{}
+	for src, dest := range volumes {
+		bindings = append(bindings, volumeMount(src, dest))
+	}
+
+	containerConfig := container.Config{
+		Tty:          true, // allocate a tty for entrypoint logs
+		Image:        image,
+		User:         fmt.Sprintf("%s:%s", user, group),
+		Cmd:          args,
+		AttachStdout: true,
+		AttachStderr: true,
+	}
+	hostConfig := container.HostConfig{
+		SecurityOpt:  []string{"seccomp=unconfined"}, // ignore seccomp
+		Binds:        bindings,
+		NetworkMode:  container.NetworkMode(net),
+		PortBindings: nat.PortMap{},
+	}
+	networkConfig := network.NetworkingConfig{}
+
+	envVars := []string{}
+	for key, val := range env {
+		envVars = append(envVars, fmt.Sprintf("%s=%s", key, val))
+	}
+
+	// pass proxy environment variables to be used by node's docker daemon
+	proxyDetails := getProxyDetails()
+	for key, val := range proxyDetails.Envs {
+		envVars = append(envVars, fmt.Sprintf("%s=%s", key, val))
+	}
+	containerConfig.Env = envVars
+
+	if usernsRemap(ctx) {
+		// We need this argument in order to make this command work
+		// in systems that have userns-remap enabled on the docker daemon
+		hostConfig.UsernsMode = "host"
+	}
+
+	// Make sure we have the image
+	err = PullDockerImage(ctx, image)
+	if err != nil {
+		return err
+	}
+
+	// Create the container using our settings
+	resp, err := docker.ContainerCreate(
+		ctx,
+		&containerConfig,
+		&hostConfig,
+		&networkConfig,
+		nil,
+		"",
+	)
+	if err != nil {
+		return err
+	}
+
+	// Read out any output from the container
+	attachOpts := types.ContainerAttachOptions{
+		Stream: true,
+		Stdin:  false,
+		Stdout: true,
+		Stderr: true,
+	}
+
+	// Attach to the container so we can capture the output
+	containerOutput, err := docker.ContainerAttach(ctx, resp.ID, attachOpts)
+	if err != nil {
+		return err
+	}
+
+	// Actually start the container
+	if err := docker.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{}); err != nil {
+		return err
+	}
+
+	outputErrors := make(chan error)
+	go func() {
+		// Send the output to the host file
+		_, err = io.Copy(output, containerOutput.Reader)
+		outputErrors <- err
+	}()
+	defer containerOutput.Close()
+
+	select {
+	case err := <-outputErrors:
+		if err != nil {
+			return err
+		}
+	case <-ctx.Done():
+		return err
```

Comment:
Should we try some kind of container cleanup when it fails here or on the wait below?

Reply:
Maybe we should bring that up in a meeting or on Slack to talk that through. My initial thought was if we are running tests and something goes wrong, we may want that container kept around to help troubleshoot what caused the failure. But maybe if we can capture enough failure information that wouldn't be necessary.

Reply:
I agree, let's delegate the responsibility about the decision to cleanup, retry or leave containers around to the caller of this methods.